pkg/controller/operators/validatingroundtripper/validating_round_tripper.go

```diff
@@ -13,23 +17,71 @@ import (
 
 type validatingRoundTripper struct {
 	delegate http.RoundTripper
+	codecs   serializer.CodecFactory
+}
+
+func (rt *validatingRoundTripper) decodeYAMLOrJSON(body io.Reader) (*unstructured.Unstructured, error) {
+	dec := yaml.NewYAMLOrJSONDecoder(body, 10)
+	unstructuredObject := &unstructured.Unstructured{}
+	if err := dec.Decode(unstructuredObject); err != nil {
+		return nil, fmt.Errorf("error decoding yaml/json object to an unstructured object: %w", err)
+	}
+	return unstructuredObject, nil
+}
+
+func (rt *validatingRoundTripper) decodeProtobuf(body io.Reader) (*unstructured.Unstructured, error) {
+	data, err := io.ReadAll(body)
+	if err != nil {
+		panic(fmt.Errorf("failed to read request body: %w", err))
+	}
+
+	decoder := rt.codecs.UniversalDeserializer()
+	obj, _, err := decoder.Decode(data, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode protobuf data: %w", err)
+	}
+
+	unstructuredObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert object to unstructured: %w", err)
+	}
+
+	return &unstructured.Unstructured{Object: unstructuredObj}, nil
+}
+
+func (rt *validatingRoundTripper) decodeRequestBody(req *http.Request) *unstructured.Unstructured {
+	b, err := req.GetBody()
+	if err != nil {
+		panic(fmt.Errorf("failed to get request body: %w", err))
+	}
+	defer b.Close()
+
+	var unstructuredObject *unstructured.Unstructured
+	switch req.Header.Get("Content-Type") {
+	case "application/vnd.kubernetes.protobuf":
+		unstructuredObject, err = rt.decodeProtobuf(b)
+	default:
+		unstructuredObject, err = rt.decodeYAMLOrJSON(b)
+	}
+
+	if err != nil {
+		panic(fmt.Errorf("failed to decode request body: %w", err))
```

Comment:
What was the determining factor between `panic` vs returning an error? Since we're dealing with network calls, it feels incorrect to panic.

Reply:
I guess the theory is that because this is a `RoundTripper` that operates on an `http.Request` that was built in memory by the same process, then there is always the expectation that we have a valid body.

However, a `RoundTripper` _could_ be used in a proxy scenario where the request is actually coming in from the network.

Reply:
I think this was originally done by SteveK with the memory optimization and resource labeling calls. I think this is specifically done on the outgoing side when creating resources. I think the idea was to ensure that all resources being created carried the labels (or die trying). I think, in the original intent, this is before the n/w call is made.